pkg/mixer/runbook.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2018 mixtool authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mixer
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"text/template"
+
+	"github.com/google/go-jsonnet"
+)
+
+const markdownTemplate = `# Alert Runbook
+{{range .Groups}}
+### {{.Name}}
+{{range .Rules}}
+##### {{.Alert}}
++ *Severity*: {{.Labels.Severity}}
++ *Message*: ` + "`" + `{{.Annotations.Message}}` + "`" + `
+
+{{.RunbookOutput}}
+{{end}}{{end}}
+`
+
+type runbookJSON struct {
+	Groups []struct {
+		Name  string `json:"name"`
+		Rules []struct {
+			Alert  string `json:"alert"`
+			Labels struct {
+				Severity string `json:"severity"`
+			} `json:"labels"`
+			Annotations struct {
+				Message string `json:"message"`
+			} `json:"annotations"`
+			RunbookOutput string `json:"runbookOutput"`
+		} `json:"rules"`
+	} `json:"groups"`
+}
+
+type RunbookOptions struct {
+	JPaths       []string
+	TemplateFile string
+}
+
+func Runbook(w io.Writer, filename string, opts RunbookOptions) error {
+	vm := jsonnet.MakeVM()
+
+	vm.Importer(&jsonnet.FileImporter{
+		JPaths: opts.JPaths,
+	})
+
+	snippetTemplate := `
+(import '%s') {
```

Comment:
let's extract the majority of this into an actual `.libsonnet` file and then just end up with:

```
(import '%s') +
(import 'mixtool-internal/runbook.libsonnet') // this is an in memory provider for the jsonnet VM
```

Reply:
Good idea!
I'm not 100% sure on what exactly you have in mind. Create this folder on disk and then use something like go-bindata (I'd prefer https://github.com/gobuffalo/packr/) to compile these files into the binary?

Reply:
yes, compiling that into the binary is what I had in mind